src/Middleware/OutputCaching/src/FormatterBinaryReader.cs

```diff
@@ -0,0 +1,197 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.OutputCaching;
+
+internal ref struct FormatterBinaryReader
+{
+    // this is effectively a cut-down re-implementation of BinaryReader
+    // from https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
```

Comment:
Nit: For ease of future archaeology, link to the commit where it was taken from rather than main.